### PR TITLE
feat(sg): send analytic events to bigquery

### PR DESCRIPTION
**Pull request by burmudar**
Reads events from the local analytics db and writes them to events table in the bigquery `sg_analytics` dataset
#### Example of published events
```
{
  "uuid": "01908369-9f55-7ed0-b8ff-79a804e4ead5",
  "user_id": "anonymous",
  "inserted_at": "2024-07-05 15:03:25.108429 UTC",
  "recorded_at": "2024-07-05 15:01:00.578734 UTC",
  "command": "sg live",
  "version": "unknown",
  "flags_and_args": "{\"args\":[\"dotcom\"],\"flags\":{\"live\":null,\"sg\":[\"disable-analytics\"]}}",
  "duration": "0-0 0 0:0:4",
  "error": "",
  "data": null,
  "metadata": "{\"cancelled\":false,\"failed\":false,\"panicked\":false,\"success\":true}"
}, {
  "uuid": "0190836b-e1e3-7b2b-8002-e5eba10f2a27",
  "user_id": "anonymous",
  "inserted_at": "2024-07-05 15:04:22.344947 UTC",
  "recorded_at": "2024-07-05 15:03:33.360211 UTC",
  "command": "sg bazel configure",
  "version": "unknown",
  "flags_and_args": "{\"args\":[],\"flags\":{\"bazel\":null,\"configure\":null,\"sg\":[\"disable-analytics\"]}}",
  "duration": "0-0 0 0:0:9",
  "error": "signal: killed",
  "data": null,
  "metadata": "{\"cancelled\":true,\"failed\":true,\"panicked\":false,\"success\":false}"
}, {
  "uuid": "0190836c-c191-78f4-93f5-d2c0da4e0873",
  "user_id": "anonymous",
  "inserted_at": "2024-07-05 15:06:04.153071 UTC",
  "recorded_at": "2024-07-05 15:05:27.525827 UTC",
  "command": "sg start",
  "version": "unknown",
  "flags_and_args": "{\"args\":[\"enterprise-bazel\"],\"flags\":{\"sg\":[\"disable-analytics\"],\"start\":null}}",
  "duration": "0-0 0 0:1:6",
  "error": "",
  "data": null,
  "metadata": "{\"cancelled\":true,\"failed\":false,\"panicked\":false,\"success\":true}"
}, {
  "uuid": "0190836e-4f48-7718-80c2-9ba82636b4ae",
  "user_id": "anonymous",
  "inserted_at": "2024-07-05 15:07:41.263882 UTC",
  "recorded_at": "2024-07-05 15:06:10.849122 UTC",
  "command": "sg cloud ephemeral list-versions",
  "version": "unknown",
  "flags_and_args": "{\"args\":[],\"flags\":{\"cloud\":null,\"ephemeral\":null,\"list-versions\":null,\"sg\":[\"disable-analytics\"]}}",
  "duration": "0-0 0 0:0:7",
  "error": "",
  "data": null,
  "metadata": "{\"cancelled\":false,\"failed\":false,\"panicked\":false,\"success\":true}"
}]	
```
#### Fields that are published
The following event attributes are puslished:
- command
- version
- flags_ang_args
- error
- recorded_at
- metadata: which contains cancelled, failed, panicked and sucess (should this possibly be stored under status instead?)

The metadata field can and _should_ be expanded.

## What is missing?
* Need to expand what is added to metadata
* Need to use and expand what is put into data

## Test plan
Tested locally and https://console.cloud.google.com/bigquery?referrer=search&project=sourcegraph-local-dev&ws=!1m10!1m4!4m3!1ssourcegraph-local-dev!2ssg_analytics!3sevents!1m4!1m3!1ssourcegraph-local-dev!2sbquxjob_761f4a78_1908370012d!3sus-central1

### CODE REVIEW

File: dev/sg/internal/analytics/sqlite.go [hunk 1]
```diff
@@ -11,6 +11,7 @@ import (
 
 type Execer interface {
 	Exec(query string, args ...interface{}) (sql.Result, error)
+	Query(stmt string, args ...interface{}) (*sql.Rows, error)
```
Comment (burmudar) [inline dev/sg/internal/analytics/sqlite.go:14]: Either this needs to go onto it's own interface `Queryer` or we need to rename this interface?

File: dev/sg/internal/analytics/analytics.go [hunk 2]
```diff
@@ -106,6 +225,55 @@ func (s analyticsStore) AddMetadata(ctx context.Context, uuid uuid.UUID, meta ma
 	}
 }
 
+func (s analyticsStore) DeleteInvocation(ctx context.Context, uuid string) {
```
Comment (jhchabran) [inline dev/sg/internal/analytics/analytics.go:228]: What about returning errors here for better testability, and then discarding those from the caller's side than silencing them here from the get go? That will also untangle the "logging" code from this along the way.

File: dev/sg/internal/analytics/background.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+		processEvents(bgOut, store(), done)
+	})
+}
+
+func StopBackgroundEventPublisher() {
+	close(done)
+}
+
+func toEvents(items []invocation) []event {
+	results := []event{}
```
Comment (jhchabran) [inline dev/sg/internal/analytics/background.go:37]: nit.nit: we know how many items we have, so why not create it with the right size from the get go?

File: dev/sg/internal/analytics/background.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	for _, i := range items {
+		ev := NewEvent(i)
+		results = append(results, *ev)
+	}
+
+	return results
+}
+
+func processEvents(bgOut *std.Output, store analyticsStore, done chan struct{}) {
+	ctx := context.Background()
```
Comment (jhchabran) [inline dev/sg/internal/analytics/background.go:47]: Shouldn't we take the ctx from the caller instead? This way we can avoid having the caller handling that part themselves with the done channel?

File: dev/sg/internal/analytics/bigquery.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+type event struct {
+	UUID         string          `json:"uuid"`
+	UserID       string          `json:"user_id"`
+	RecordedAt   time.Time       `json:"recorded_at"`
+	Command      string          `json:"command"`
+	Version      string          `json:"version"`
+	FlagsAndArgs json.RawMessage `json:"flags_and_args,omitempty"`
+	Duration     time.Duration   `json:"duration,omitempty"`
+	Error        string          `json:"error,omitempty"`
+	Data         json.RawMessage `json:"data,omitempty"`
```
Comment (jhchabran) [inline dev/sg/internal/analytics/bigquery.go:27]: 💯

File: dev/sg/internal/analytics/bigquery.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	Metadata     json.RawMessage `json:"metadata,omitempty"`
+}
+
+func (e event) Save() (map[string]bigquery.Value, string, error) {
```
Comment (jhchabran) [inline dev/sg/internal/analytics/bigquery.go:31]: unused?

Comment (burmudar) [reply dev/sg/internal/analytics/bigquery.go:31]: It's implements the Saver interface of bigquery which is used by `Table.Inserter` - I'll add a comment here

File: dev/sg/internal/analytics/analytics.go [hunk 7]
```diff
@@ -27,7 +27,125 @@ const (
 )
 
 type invocation struct {
-	uuid uuid.UUID
+	uuid     uuid.UUID
+	metadata map[string]any
+}
+
+func (i invocation) GetStartTime() *time.Time {
```
[the remark this answers is not in the document]
Comment (burmudar) [reply dev/sg/internal/analytics/analytics.go:34]: This is an excellent point 🙏🏼

File: dev/sg/internal/analytics/analytics.go (hunk 2 above)
Comment (burmudar) [reply dev/sg/internal/analytics/analytics.go:228]: Yeah I agree. Currently it's done for consistency with the other methods, but I'll update it

Comment (burmudar) [reply dev/sg/internal/analytics/analytics.go]: @jhchabran I've updated the code here to return the err and at the top level where analytics get registered on all commands the error is reported. Below is the output from my testing
```
# database is not readable nor writeable

web installed
⠧  6/13 commands installed  ████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████                                                                                                                                                            46%
^C⚠️ Interrupt received, executing hook groups for graceful shutdown...
⚠️ Executing 1 'cleanup' hooks for graceful shutdown...
⚠️ Executing 6 'general' hooks for for graceful shutdown...
⠧  6/13 commands installed  ████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████                                                                                                                                                            46%
⚠️ failed to persist analytics cancel event: analytics database not initialized
exit status 1

sourcegraph on  wb/sg/analytics-send-events [$!?⇡] via 🐹 v1.22.4 via ❄️  impure (sourcegraph-dev-env) took 40s
❯ SG_DISABLE_ANALYTICS=0 SG_ANALYTICS_DEBUG=1 go run ./dev/sg cloud eph status
⚠️ Running sg with a dev build, following flags have different default value unless explictly set: skip-auto-update, disable-analytics
⚠️ Failed to create sg analytics store: unable to open database file: out of memory (14)
❌ getting status of "wb-sg-analytics-send-events" failed
⚠️ failed to persist analytics cancel event: analytics database not initialized
❌ instance not found
exit status 1

# database readable and writeable again
sourcegraph on  wb/sg/analytics-send-events [$!?⇡] via 🐹 v1.22.4 via ❄️  impure (sourcegraph-dev-env) took 18s
❯ SG_DISABLE_ANALYTICS=0 SG_ANALYTICS_DEBUG=1 go run ./dev/sg cloud eph status
⚠️ Running sg with a dev build, following flags have different default value unless explictly set: skip-auto-update, disable-analytics
❌ getting status of "wb-sg-analytics-send-events" failed
❌ instance not found
exit status 1
```

File: dev/sg/internal/analytics/analytics.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (burmudar) [reply dev/sg/internal/analytics/analytics.go:37]: Still have to look into this

**Review by jhchabran — APPROVED** [review]

**Pull request merged**